[FEATURE] Write new chapter on project testing (#4509)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/Testing/ProjectTesting.rst

Project testing
===============

..  contents::

.. _testing-projects-differences:

Differences between project and extension testing
=================================================

**Projects** usually needs to support only one PHP version, Database vendor
and version and TYPO3 core version

Version raises for upgrades are usually prepared on a branch and changed
instead of parallel execution.

Project may have different places for tests

*   local path extension tests `packages/*/Tests/*`
*   global (root) tests `Tests/*`

The Core mono repository is basically a project setup, having local path
extensions in `typo3/sysexts/*` instead of the more known and lived `packages/*`
project folder structure.

> .. _testing-projects-structure:

Reviewer: ```suggestion
.. _testing-project-structure:
```

Project structure
=================

We assume a project structure similar to `tf-basics-project <https://github.com/sbuerk/tf-basics-project>`__
here. If you are using another structure, you have to adjust some scripts.

*   :ref:`TYPO3 installation with Composer <t3start:install>` and default paths
*   :ref:`DDEV <t3start:installation-ddev-tutorial>` is used for local development
*   Local extensions and the site package are situated in :path:`packages`

..  directory-tree::
    :caption: Initial directory tree in version control
    :level: 1

    *   :path:`.ddev`
    *   :path:`config`
    *   :path:`packages`
    *   :file:`composer.json`
    *   :file:`composer.lock`

The :file:`composer.json` looks like this:

..  literalinclude:: _ProjectTesting/_composer.json
    :caption: Example project composer.json before testing

> .. _testing-projects-installation:

Reviewer: ```suggestion
..  _testing-projects-installation:
```

Install testing dependencies
============================

As a bare minimum it is suggested to use

*   One coding style fixer for PHP, for example :composer:`friendsofphp/php-cs-fixer`
*   One static code analyzer for PHP, for example :composer:`phpstan/phpstan`

Depending on the complexity of your project you might need:

*   :composer:`phpunit/phpunit`, if there is any PHP code of a complexity that
    should be tested.
*   Testing of scss, TypeScript or JavaScript (not covered here)
*   Linting of YAML, XML, TypoScript (not covered here)
*   :ref:`Writing acceptance tests <t3coreapi:testing-writing-acceptance>`

You can install all these tools as development dependencies. They will then not
be installed on your production system when Composer is executed with option
`--no-dev` during deployment:

..  code-block:: bash
    :caption: Composer installation **during deployment**

    composer install --no-dev

> For TYPO3 project you can use the package :composer:`typo3/coding-standards`

Author: ```suggestion
..  todo: Link to a dedicated php-cs-fixer section once it exists

For TYPO3 project you can use the package :composer:`typo3/coding-standards`
```

which bundles :composer:`friendsofphp/php-cs-fixer`, :composer:`phpstan/phpstan`,
:composer:`phpunit/phpunit` and some other helpful tools and comes with useful

> configurations for these.

Reviewer: @sbuerk I remember a talk with you, that you do not install that package. Correct me, if I quote you wrong: "That package would be nice, if there would be someone keeping the work up-2-date. Currenty, you can just install php-cs-fixer as dependency, take the config files from core and you're done. So, no need to install that package"

Author: Yes you can but then You use the advantage of a common package

Reviewer 2: I would also rather like to encourage people to not use a bundled install, but maintain the configs here, so that they understand each tool and are able to react if a "meta package" isn't used.

We can hint at that package for inspiration or an alternate approach but maybe not encourage to use that one?

Reviewer 3: > @sbuerk I remember a talk with you, that you do not install that package. Correct me, if I quote you wrong: "That package would be nice, if there would be someone keeping the work up-2-date. Currenty, you can just install php-cs-fixer as dependency, take the config files from core and you're done. So, no need to install that package"

That as a purly personal preference and point of view, without the intention to blame anybody. It's till a legit packagit to be used if it suits your/someones need. My personal view here should not be taken as a general recommandation - and I thought I had made that clear.

Reviewer 3: Maybe having a dedicated section somewhere related to `code style linting/fixing` and providing/explaining usual used tools - raw or with the `typo3/coding-standards` package. Developers/Teams can than decide on their preferences what to use and having the guide at hand - and link from here.

Author: Such a section is a good idea, but as long as we don't have one using the one from `typo3/coding-standards` is certainly better then none. And as the topic is opinionated anyway I would stick with. Use whatever you want, for example you can use..

Author: Once the section exists we can link it from here

Author: I added a todo for that.

..  code-block:: bash
    :caption: Require development dependencies

    composer req --dev typo3/coding-standards

If you want to do Unit or Functional tests on project level you also need the
TYPO3 testing framework:

..  code-block:: bash
    :caption: Require development dependencies

    composer req --dev typo3/coding-standards typo3/testing-framework

> .. _testing-projects-configuration:

Reviewer: Sure, not a must have, but first title says "Differences between project and extension setting". So, I prefer using "project" here:

```suggestion
.. _testing-project-configuration:
```

Test configuration on project level
===================================

We suggest to keep all project level test configuration in a common place that
can be excluded from deployment. The Core uses a folder called :path:`Build` with
one folder per test-type and we will follow that scheme here. If you put
the configuration in other directories, adjust your configuration files
accordingly.

> .. _testing-projects-configuration-cs:

Reviewer: ```suggestion
..  _testing-projects-configuration-cs:
```

Code style tests and fixing
---------------------------

> :composer:`typo3/testing-framework` comes with a predefined configuration for

Reviewer 3: testing-framework cs-fixer configuration is not meant as template to be used. It's purly for that repository. We do not maintain it in regards for other projects or extensions.

Either move coding style to a dedicated section describing all possibilites side by side - or at least stay with one way in this document. Above, the typo3/coding-standards are required - but the config from testing framework should be used ? That does not make much sence. Or do I miss here something ?

Author: Sorry, mixed them up, fixed it.

> :composer:`typo3/testing-framework` comes with a predefined configuration for

Author: ```suggestion
:composer:`typo3/coding-standards` comes with a predefined configuration for
```

:composer:`friendsofphp/php-cs-fixer`. You can override rules as needed in
your own configuration:

..  literalinclude:: _ProjectTesting/_php-cs-fixer.dist.php
    :caption: Build/php-cs-fixer/.php-cs-fixer.dist.php

> It is recommended to also copy the :file:`.editorconfig` from the testing

Reviewer: Very good. Thank you for taken that into account

> framework into your main directory so that your IDE applies the same formating

Reviewer: ```suggestion
framework into your main directory so that your IDE applies the same formatting
```

as the php-cs-fixer.

.. _testing-projects-configuration-phpstan:

PHPstan - Static PHP analysis
-----------------------------

When configuring PHPstan the various places in which PHP files can be found
should be taken into consideration:

..  literalinclude:: _ProjectTesting/_phpstan.neon
    :caption: Build/phpstan/phpstan.neon

It also makes sense to exclude the :file:`ext_emconf.php` and any
:path:`node_modules` directory.

> .. _testing-projects-configuration-phpunit:

Reviewer: ```suggestion
..  _testing-projects-configuration-phpunit:
```

Unit and Functional test configuration
--------------------------------------

See the chapters :ref:`Unit testing <testing-writing-unit>` and
:ref:`Functional testing <testing-writing-functional>`.

> .. _testing-projects-execution:

Reviewer: ```suggestion
..  _testing-projects-execution:
```

Running the tests locally
=========================

The tests can be run via PHP on your local machine or with DDEV.

> .. _testing-projects-execution-cs:

Reviewer: ```suggestion
..  _testing-projects-execution-cs:
```

> Run the PHP cs fixer

Reviewer: In various parts of the doc you write "php-cs-fixer". You can keep it, but that writing is different than all other mentions of that package

Reviewer 2: I think here we also need a unique spelling, I think "PHP-CS-Fixer" is the right way

Author: It is not in our Glossary yet: https://docs.typo3.org/m/typo3/docs-how-to-document/main/en-us/GeneralConventions/Glossary.html would one of you like to make a descision and make a PR there? then we can do search and replace all over.

> Run the PHP cs fixer

Author: ```suggestion
Run the php-cs-fixer
```

--------------------

To run the PHP cs fixer you need to configure the path to the configuration
file:

..  code-block:: bash

>     vendor/bin/php-cs-fixer fix  --config=Build/php-cs-fixer/.php-cs-fixer.dist.php

Reviewer: ```suggestion
    vendor/bin/php-cs-fixer fix --config=Build/php-cs-fixer/.php-cs-fixer.dist.php
```

.. _testing-projects-execution-phpstan:

Run PHPstan
-----------

..  code-block:: bash

    vendor/bin/phpstan --configuration=Build/phpstan/phpstan.neon

Regenerate the baseline:

..  code-block:: bash

    vendor/bin/phpstan \
        --configuration=Build/phpstan/phpstan.neon \
        --generate-baseline=Build/phpstan/phpstan-baseline.neon

> .. _testing-projects-execution-phpunit:

Reviewer: ```suggestion
..  _testing-projects-execution-phpunit:
```

Run Unit tests
--------------

As Unit tests need no database or other dependencies you can run them directly
on your host system or DDEV:

..  code-block:: bash

    vendor/bin/phpunit \
        -c Build/phpunit/UnitTests.xml

> .. _testing-projects-execution-functional-sqlite:

Reviewer: ```suggestion
..  _testing-projects-execution-functional-sqlite:
```

extensions in `typo3/sysexts/*` instead of the more known and lived `packages/*`
project folder structure.

> .. _testing-project-structure:

Reviewer: Maybe it would be good to keep the old link. No need to create redirects or whatever:
```suggestion
.. _testing-projects-structure:
```

    composer req --dev typo3/coding-standards typo3/testing-framework

> .. _testing-project-configuration:

Reviewer: Maybe it would be good to keep the old link. No need to create redirects or whatever:
```suggestion
.. _testing-projects-configuration:
```

It also makes sense to exclude the :file:`ext_emconf.php` and any
:path:`node_modules` directory.

..  _testing-projects-configuration-phpunit:

Unit and Functional test configuration
--------------------------------------

See the chapters :ref:`Unit testing <testing-writing-unit>` and
:ref:`Functional testing <testing-writing-functional>`.

..  _testing-projects-execution:

Running the tests locally
=========================

The tests can be run via PHP on your local machine or with DDEV.

..  _testing-projects-execution-cs:

Run the php-cs-fixer
--------------------

> To run the PHP cs fixer you need to configure the path to the configuration

Reviewer: ```suggestion
To run the php-cs-fixer you need to configure the path to the configuration
```
